### PR TITLE
[RFC] Implement task monad

### PR DESCRIPTION
**Pull request by alex-lairan**
Add the Task monad to the monads shard.

# What is `Task` monad ?

The `Task` monad permit to run asynchronous action and chain it like a monad.

# Why Task don't inherit from `Monads::Monad(T)` ?

The Task monad doesn't have the `fmap` method.
Also, it doesn't realy behave like a monad, it could be converted to a monad with `to_either` or `to_maybe`.

### CODE REVIEW

File: src/monads/task.cr [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    end
+
+    @final_value : T?
+    @final_error : Exception?
+    @fiber : Fiber
+
+    @kind = Message::None
+
+    def initialize(proc)
+      @channel_message = channel_message = Channel::Buffered(Message).new
```
Comment (alex-lairan) [inline src/monads/task.cr]: This is a workaround to know if a task is finished.

File: src/monads/task.cr [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+          channel_message.send Message::Value
+          channel_value.send value
+        rescue exception
+          channel_message.send Message::Error
+          channel_error.send exception
+        end
+      end
+    end
+
+    def bind(lambda : T -> U) : U forall U
```
Comment (alex-lairan) [inline src/monads/task.cr]: This don't realy behave like others `bind`.

Another name could be used.

File: spec/monads/task_spec.cr [hunk 3]
```diff
@@ -0,0 +1,43 @@
+require "../spec_helper"
+
+class Exception
```
Comment (alex-lairan) [inline spec/monads/task_spec.cr]: An exception cannot be compared.

But the `Either` monad will have an Exception.

I think we could create a class who inherit from an Either for this use case.

File: src/monads/task.cr [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+      @channel_error = channel_error = Channel::Buffered(Exception).new
+      @final_value = nil
+      @final_error = nil
+
+      spawn do
+        begin
+          value = proc.call
+          channel_message.send Message::Value
+          channel_value.send value
+        rescue exception
```
Comment (jwoertink) [inline src/monads/task.cr]: Should this be `rescue Exception`?

Comment (jwoertink) [reply src/monads/task.cr]: oh, maybe not. Never mind 😅 .

**Comment by alex-lairan** [comment]
@moba1 can you look at this code ?

File: src/monads/task.cr [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+    def bind(lambda : T -> U) : U forall U
+      lambda.call(receive)
+    end
+
+    def receive : T
+      case @kind
+      when Message::None
+        @kind = @channel_message.receive
+        receive
```
Comment (moba1) [inline src/monads/task.cr]: Here recusive call is likely to be stack overflow.

Comment (alex-lairan) [reply src/monads/task.cr]: No because it will be called only one time.

**Comment by moba1** [comment]
LGTM

**Comment by alex-lairan** [comment]
I disagree, there is the Exception problem, I try to fix it now ! :)

**Comment by alex-lairan** [comment]
I made a solution :) 

Take a look

**Comment by moba1** [comment]
🙆‍♂

**Pull request merged**